File: src/main/java/org/tindalos/principle/domain/core/PackageReference.java
package org.tindalos.principle.domain.core;
import org.apache.commons.lang3.builder.HashCodeBuilder;

/**
 * Represents a reference to a Java package by its fully qualified name.
 * Provides hierarchical package navigation and comparison operations.
 *
 * @param name the fully qualified package name
 */
public record PackageReference(String name) implements Comparable<PackageReference> {

    public boolean startsWith(String str) {
        return name.startsWith(str);
    }

    public String createChild(String relativeNameOfDirectSubPackage) {
        return name + "." + relativeNameOfDirectSubPackage;
    }

    public boolean isDirectParentOf(PackageReference reference) {
        if (this.equals(reference)) {
            return false;
        }
        return !reference.relativeNameTo(this).contains(".");
    }

    public boolean isNotAnAncestorOf(PackageReference reference) {
        return !reference.pointsInside(this);
    }

    public PackageReference child(String relativeName) {
        return new PackageReference(name + "." + relativeName);
    }

    public boolean pointsInside(PackageReference reference) {
        return startsWith(reference.name + ".");
    }

    public boolean pointsToThatOrInside(PackageReference reference) {
        return this.equals(reference) || pointsInside(reference);
    }

    public boolean oneContainsAnother(PackageReference that) {
        return pointsToThatOrInside(that) || that.pointsToThatOrInside(this);
    }

    public boolean isDescendantOf(PackageReference reference) {
        return startsWith(reference.name + ".");
    }

    public String relativeNameTo(PackageReference reference) {
        if (!this.name.startsWith(reference.name + ".")) {
            throw new IllegalArgumentException(this + " is not under " + reference);
        }
        return name.replaceFirst(reference.name + "\\.", "");
    }

    public String firstPartOfRelativeNameTo(PackageReference reference) {
        return relativeNameTo(reference).split("\\.", 2)[0];
    }


    @Override
    public int compareTo(PackageReference that) {
        return name.compareTo(that.name);
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(name).hashCode();
    }
}

File: src/main/scala/org/tindalos/principle/domain/core/PackageReference.scala
package org.tindalos.principle.domain.core

import org.apache.commons.lang3.builder.HashCodeBuilder

class PackageReference(val name: String) extends Comparable[PackageReference] {
  
  def startsWith(str: String) = name.startsWith(str)

  def createChild(relativeNameOfDirectSubPackage: String) = name + "." + relativeNameOfDirectSubPackage

  def isDirectParentOf(reference: PackageReference) =
    if (this.equals(reference)) false
    else !reference.relativeNameTo(this).contains(".")

  def isNotAnAncestorOf(reference: PackageReference) = !reference.pointsInside(this)

  def child(relativeName: String) = new PackageReference(name + "." + relativeName)

  def pointsInside(reference: PackageReference) = startsWith(reference.name + ".")

  def pointsToThatOrInside(reference: PackageReference) = this.equals(reference) || pointsInside(reference)

  def oneContainsAnother(that: PackageReference) = pointsToThatOrInside(that) || that.pointsToThatOrInside(this)

  def isDescendantOf(reference: PackageReference) = startsWith(reference.name + ".")

  def relativeNameTo(reference: PackageReference) =
    if (!this.name.startsWith(reference.name + ".")) {
      throw new IllegalArgumentException(this + " is not under " + reference);
    } else
      name.replaceFirst(reference.name + ".", "")

  def firstPartOfRelativeNameTo(reference: PackageReference) = relativeNameTo(reference).split("\\.", 2).head

  override def toString() = name
  
  override def equals(other:Any) = if (!other.isInstanceOf[PackageReference]) false else other.asInstanceOf[PackageReference].name.equals(this.name)
  
  override def hashCode() = new HashCodeBuilder().append(name).hashCode()

  override def compareTo(that: PackageReference) = name.compareTo(that.name)
}